### PR TITLE
Extend logs for warnings

### PR DESCRIPTION
**Pull request by xsebek**
* split logs into a separate module
* make some logs warnings and fatal (to be continued...)
* prepare failure logging infrastructure for TUI

This should be merged before #954 so it is not squashed.

### CODE REVIEW

File: src/Swarm/TUI/Controller.hs [hunk 1]
```diff
@@ -109,8 +109,8 @@ handleEvent = \case
   AppEvent (UpstreamVersion ev) -> do
     let logReleaseEvent l e = runtimeState . eventLog %= logEvent l ("Release", -7) (T.pack $ show e)
     case ev of
-      Left e@(FailedReleaseQuery _e) -> logReleaseEvent ErrorTrace e
-      Left e -> logReleaseEvent Said e
```
Comment (xsebek) [inline src/Swarm/TUI/Controller.hs:113]: @kostmo @byorgey FYI: this used to create nice yellow warning logs by abusing the fact that _messages said by the robot with ID `-7` happen to be in yellow._ :sweat_smile:

File: src/Swarm/Game/Log.hs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    ErrorTrace ErrorLevel
+  deriving (Show, Eq, Ord, Generic, FromJSON, ToJSON)
+
+-- | An entry in a robot's log.
+data LogEntry = LogEntry
+  { _leTime :: Integer
+  -- ^ The time at which the entry was created.
+  --   Note that this is the first field we sort on.
+  , _leSaid :: LogSource
+  -- ^ Whether this log records a said message.
```
Comment (byorgey) [inline src/Swarm/Game/Log.hs:62]: Maybe we should change the name and comment of this field (`_leSource`, perhaps?). The current name and comment make it sound like it is a `Bool`.

File: src/Swarm/Game/Log.hs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  , _leRobotName :: Text
+  -- ^ The name of the robot that generated the entry.
+  , _leRobotID :: Int
+  -- ^ The ID of the robot that generated the entry.
+  , _leLocation :: Location
+  -- ^ Location of the robot at log entry creation.
+  , _leText :: Text
```
Comment (byorgey) [inline src/Swarm/Game/Log.hs:69]: At https://github.com/swarm-game/swarm/pull/959#issuecomment-1372626765 we discussed maybe storing something more structured here than just a `Text`, which I think is worth exploring.  But we should probably leave that to a different PR.

File: src/Swarm/TUI/Model/Failure.hs [hunk 4]
```diff
@@ -0,0 +1,51 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+-- |
+-- Module      :  Swarm.Game.Log
```
Comment (byorgey) [inline src/Swarm/TUI/Model/Failure.hs:4]: Fix this module name

File: src/Swarm/TUI/Model/Failure.hs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+-- Copyright   :  Ondřej Šebek
+-- Maintainer  :  ondras98@icloud.com
+--
+-- SPDX-License-Identifier: BSD-3-Clause
+--
+-- A data type to represent sustem failures.
```
Comment (byorgey) [inline src/Swarm/TUI/Model/Failure.hs:10]:
```suggestion
-- A data type to represent system failures.
```

**Review by byorgey — APPROVED** [review]
Looks good!

File: src/Swarm/Game/Log.hs (hunk 3 above)
Comment (xsebek) [reply src/Swarm/Game/Log.hs:69]: I would like to split a separate type for system logs which should also use structured errors.

I think that will make the design clear, now it's all tangled and adding anything to it would be a mistake IMO.

File: src/Swarm/TUI/Model/Failure.hs (hunk 4 above)
Comment (xsebek) [reply src/Swarm/TUI/Model/Failure.hs:4]:
```suggestion
-- Module      :  Swarm.Game.Failure
```

Comment (byorgey) [reply src/Swarm/TUI/Model/Failure.hs]: But isn't it `TUI.Model.Failure`?

File: src/Swarm/Game/Log.hs (hunk 2 above)
Comment (xsebek) [reply src/Swarm/Game/Log.hs]: I am sure it was a `Bool` at some point. :slightly_smiling_face: Now it's fixed.

File: src/Swarm/Game/Log.hs [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    Said
+  | -- | Log produced by 'Swarm.Language.Syntax.Log'
+    Logged
+  | -- | Log produced by an exception or system.
+    ErrorTrace ErrorLevel
+  deriving (Show, Eq, Ord, Generic, FromJSON, ToJSON)
+
+-- | An entry in a robot's log.
+data LogEntry = LogEntry
+  { _leTime :: Integer
```
Comment (kostmo) [inline src/Swarm/Game/Log.hs:58]: These fields sound [French](https://youtu.be/kCpjgl2baLs?t=45) :stuck_out_tongue_winking_eye:
